File: hahlw/dp/84.柱状图中最大的矩形.cpp
/*
 * @lc app=leetcode.cn id=84 lang=cpp
 *
 * [84] 柱状图中最大的矩形
 */
class Solution {
public:
    //单调栈 
    /*
    遇到比栈顶元素大的元素则入栈，反之出栈，并计算当前面积。
    */
    int largestRectangleArea(vector<int>& heights) {
         stack<int>stk;
         int res = 0;
         heights.push_back(0);
         for(int i = 0;i<heights.size();i++){
            if(stk.empty()||heights[stk.top()]<heights[i]){
                stk.push(i);
            }
            else{
                int cur=stk.top();stk.pop();
                res=max(res,heights[cur]*(stk.empty()?i:(i-stk.top()-1)));
                i--;
            }
         }
         return res;
    }

    //Pruning optimize,网上的办法。
    //每次找到一个局部峰值，然后往前遍历，计算面积，更新最大面积。
     int largestRectangleArea_po(vector<int>& heights){
         int res = 0;
         for(int i =0;i<heights.size();i++){
            
            if(i+1<heights.size()&&heights[i+1]>=heights[i])continue;

            int minH=heights[i];

            for(int j=i;j>=0;j--){
                minH=min(minH,heights[j]);
                int area=minH*(i-j+1);
                res=max(res,area);
            }

         }
         return res;
     }

};

File: Juniper/dynamic programming/84.柱状图中最大的矩形.py
#
# @lc app=leetcode.cn id=84 lang=python3
#
# [84] 柱状图中最大的矩形
#
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        # lens = len(heights)
        # rect  = 0
        # for i in range(lens):
        #     left = i
        #     right = i 
        #     while left >0 and heights[left-1]>=heights[i]:
        #         left -=1 
        #     while right < lens-1 and heights[right+1]>=heights[i]:
        #         right +=1
        #     rect = max(rect, (right-left+1)*heights[i])
        # return rect

        if not heights:
            return 0
        lens = len(heights)
        left = [0 for i in range(lens)]
        right = [0 for i in range(lens)]
        left[0] = -1
        right[-1] = lens
        for i in range(1,lens):
            temp = i-1
            while temp >=0 and  heights[temp] >= heights[i] :
                temp = left[temp]
            left[i] = temp
            temp = lens-1-i+1
            while temp < lens and heights[temp] >= heights[lens-1-i]:
                temp = right[temp]
            right[lens-1-i] = temp
        rect = 0
        for i in range(lens):
            rect = max(rect, (right[i]-left[i]-1)*heights[i])
        return rect 


            



